File: packages/protocol/src/path.ts
export type PathErrorCode =
  | "empty_path"
  | "path_too_long"
  | "non_nfc"
  | "absolute_path"
  | "trailing_slash"
  | "empty_segment"
  | "dot_segment"
  | "segment_too_long"
  | "control_character"
  | "forbidden_character"
  | "trailing_dot_or_space"
  | "windows_reserved_name";

export class PathError extends Error {
  constructor(
    readonly code: PathErrorCode,
    message: string,
  ) {
    super(message);
    this.name = "PathError";
  }
}

const textEncoder = new TextEncoder();
const windowsReserved = /^(?:CON|PRN|AUX|NUL|COM[1-9]|LPT[1-9])$/;

function reject(code: PathErrorCode, message: string): never {
  throw new PathError(code, message);
}

/** Validates one path against `edgefossil-path-v0`. */
export function validatePath(path: string): void {
  const encodedLength = textEncoder.encode(path).length;
  if (encodedLength === 0) reject("empty_path", "path must not be empty");
  if (encodedLength > 4096)
    reject("path_too_long", "path exceeds 4096 UTF-8 bytes");
  if (path.normalize("NFC") !== path)
    reject("non_nfc", "path must already be Unicode NFC");
  if (path.startsWith("/"))
    reject("absolute_path", "absolute paths are forbidden");
  if (path.endsWith("/"))
    reject("trailing_slash", "path must not end with a separator");

  for (const segment of path.split("/")) {
    if (segment.length === 0)
      reject("empty_segment", "empty path segments are forbidden");
    if (segment === "." || segment === "..")
      reject("dot_segment", "dot path segments are forbidden");
    if (textEncoder.encode(segment).length > 255) {
      reject("segment_too_long", "path segment exceeds 255 UTF-8 bytes");
    }
    for (const character of segment) {
      const codePoint = character.codePointAt(0)!;
      if (codePoint <= 0x1f || codePoint === 0x7f) {
        reject("control_character", "control characters are forbidden");
      }
      if ('<>:"\\|?*'.includes(character)) {
        reject(
          "forbidden_character",
          "platform-forbidden ASCII characters are forbidden",
        );
      }
    }
    if (segment.endsWith(".") || segment.endsWith(" ")) {
      reject(
        "trailing_dot_or_space",
        "segments must not end with dot or space",
      );
    }
    const deviceStem = segment.split(".", 1)[0]!.toUpperCase();
    if (windowsReserved.test(deviceStem)) {
      reject("windows_reserved_name", "Windows device names are forbidden");
    }
  }
}

File: crates/ef-format/src/path.rs
//! Portable repository-path validation.

use std::{error::Error, fmt};

use unicode_normalization::UnicodeNormalization;

/// Stable machine-readable path validation failure.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum PathErrorCode {
    EmptyPath,
    PathTooLong,
    NonNfc,
    AbsolutePath,
    TrailingSlash,
    EmptySegment,
    DotSegment,
    SegmentTooLong,
    ControlCharacter,
    ForbiddenCharacter,
    TrailingDotOrSpace,
    WindowsReservedName,
}

/// A path rejected by `edgefossil-path-v0`.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct PathError {
    /// Stable failure category.
    pub code: PathErrorCode,
    message: &'static str,
}

impl PathError {
    const fn new(code: PathErrorCode, message: &'static str) -> Self {
        Self { code, message }
    }
}

impl fmt::Display for PathError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(self.message)
    }
}

impl Error for PathError {}

fn is_windows_reserved(segment: &str) -> bool {
    let stem = segment
        .split('.')
        .next()
        .expect("split always has one item");
    let upper = stem.to_ascii_uppercase();
    matches!(upper.as_str(), "CON" | "PRN" | "AUX" | "NUL")
        || upper
            .strip_prefix("COM")
            .or_else(|| upper.strip_prefix("LPT"))
            .is_some_and(|suffix| {
                matches!(suffix, "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9")
            })
}

/// Validates one path against `edgefossil-path-v0`.
///
/// # Errors
///
/// Returns the first failure in the profile's deterministic validation order.
pub fn validate_path(path: &str) -> Result<(), PathError> {
    if path.is_empty() {
        return Err(PathError::new(
            PathErrorCode::EmptyPath,
            "path must not be empty",
        ));
    }
    if path.len() > 4096 {
        return Err(PathError::new(
            PathErrorCode::PathTooLong,
            "path exceeds 4096 UTF-8 bytes",
        ));
    }
    if !path.nfc().eq(path.chars()) {
        return Err(PathError::new(
            PathErrorCode::NonNfc,
            "path must already be Unicode NFC",
        ));
    }
    if path.starts_with('/') {
        return Err(PathError::new(
            PathErrorCode::AbsolutePath,
            "absolute paths are forbidden",
        ));
    }
    if path.ends_with('/') {
        return Err(PathError::new(
            PathErrorCode::TrailingSlash,
            "path must not end with a separator",
        ));
    }

    for segment in path.split('/') {
        if segment.is_empty() {
            return Err(PathError::new(
                PathErrorCode::EmptySegment,
                "empty path segments are forbidden",
            ));
        }
        if matches!(segment, "." | "..") {
            return Err(PathError::new(
                PathErrorCode::DotSegment,
                "dot path segments are forbidden",
            ));
        }
        if segment.len() > 255 {
            return Err(PathError::new(
                PathErrorCode::SegmentTooLong,
                "path segment exceeds 255 UTF-8 bytes",
            ));
        }
        for character in segment.chars() {
            if character <= '\u{1f}' || character == '\u{7f}' {
                return Err(PathError::new(
                    PathErrorCode::ControlCharacter,
                    "control characters are forbidden",
                ));
            }
            if matches!(character, '<' | '>' | ':' | '"' | '\\' | '|' | '?' | '*') {
                return Err(PathError::new(
                    PathErrorCode::ForbiddenCharacter,
                    "platform-forbidden ASCII characters are forbidden",
                ));
            }
        }
        if segment.ends_with(['.', ' ']) {
            return Err(PathError::new(
                PathErrorCode::TrailingDotOrSpace,
                "segments must not end with dot or space",
            ));
        }
        if is_windows_reserved(segment) {
            return Err(PathError::new(
                PathErrorCode::WindowsReservedName,
                "Windows device names are forbidden",
            ));
        }
    }
    Ok(())
}
